src/etwlogger_driver.c

```diff
@@ -113,26 +113,98 @@ static char* lastErrorToString(DWORD lastError)
 }
 
 
-static int isETWLoggerInit = 0;
+static volatile LONG isETWLoggerInit = 0;
 
 static void lazyRegisterEventProvider(void)
 {
     /*lazily init the logger*/
-    if (isETWLoggerInit == 0)
+    if (InterlockedCompareExchange(&isETWLoggerInit, 1, 0)==0)
     {
         ULONG event_register_result = EventRegisterMicrosoft_ServiceBus();
         if (event_register_result != ERROR_SUCCESS)
         {
             /*go back to printf, maybe somebody notices*/
             (void)printf("failure in EventRegisterMicrosoft_ServiceBus: %lu\r\n", event_register_result);
+            (void)InterlockedExchange(&isETWLoggerInit, 0); /*allow a reattempt to register*/
         }
         else
         {
             /*should check if the provider is registered with the system... maybe at a later time*/
-            isETWLoggerInit = 1; /*and stays 1 until the process exits*/ /*sorry, no graceful exit with EventUnregisterMicrosoft_ServiceBus*/
+            /*isETWLoggerInit stays 1 until the process exits*/ /*sorry, no graceful exit with EventUnregisterMicrosoft_ServiceBus*/
             LogInfo("EventRegisterMicrosoft_ServiceBus success"); /*self logging that the log service has started*/
         }
     }
+    else
+    {
+        /*do nothing, already registered/attempted*/
+    }
+}
+
+/*the below interlocked variable initially set to 0*/
+/* At the first error returned by EventWriteLogInfoEvent or by EventWriteLogErrorEvent a printf will be performed and the variable will be switched to 1. */
+/* while the variable is set to "1" no fallback printf will be performed */
+/* at the first non-erroneous EventWriteLogInfoEvent or EventWriteLogErrorEvent the printf fallback will be re-enabled */
+static volatile LONG stopPrintfing = 0;
```

Comment:
>stopPrintfing [](start = 21, length = 13)

/poetry This is just a matter of preference. I would normally have liked the variable name to represent what it enables (printfing_enabled) not the action (stop). But this is so small that maybe I should not even made the comment :-).

Reply:
renamed to "doPrintf" and changed the logic to positive

---
In reply to: [319621136](https://github.com/Azure/azure-c-shared-utility/pull/367#discussion_r319621136) [](ancestors = 319621136)